src/sched/mod.rs

```diff
@@ -59,27 +134,136 @@ pub fn spawn_kernel_work(fut: impl Future<Output = ()> + 'static + Send) {
         .put_kernel_work(Box::pin(fut));
 }
 
-/// Insert the given task onto *this* CPUs runqueue.
+fn get_next_cpu() -> CpuId {
+    static NEXT_CPU: AtomicUsize = AtomicUsize::new(0);
+
+    let cpu_count = ArchImpl::cpu_count();
+    let cpu_id = NEXT_CPU.fetch_add(1, core::sync::atomic::Ordering::Relaxed) % cpu_count;
+
+    CpuId(cpu_id)
+}
+
+/// Insert the given task onto a CPU's run queue.
 pub fn insert_task(task: Arc<Task>) {
-    SCHED_STATE
-        .borrow_mut()
-        .run_queue
-        .insert(task.descriptor(), task);
+    SCHED_STATE.borrow_mut().add_task(task);
+}
+
+pub fn insert_task_cross_cpu(task: Arc<Task>) {
+    let cpu = get_next_cpu();
+    message_cpu(cpu.value(), Message::PutTask(task)).expect("Failed to send task to CPU");
```

Comment:
Could slightly optimise here to use `insert_task` if we're inserting to the current CPU.